gratipay/utils/__init__.py

```diff
@@ -132,6 +133,46 @@ def get_team(state):
     return team
 
 
+def b64encode_s(s):
+    prefix = b''
+    if not isinstance(s, bytes):
+        s = s.encode('utf8')
+    else:
+        # Check whether the string is binary or already utf8
+        try:
+            s.decode('utf8')
+        except UnicodeError:
+            prefix = b'.'
+    return prefix + b64encode(s, b'-_').replace(b'=', b'~')
+
+
+def b64decode_s(s, **kw):
+    def error():
+        if 'default' in kw:
+            return kw['default']
+        raise Response(400, "invalid base64 input")
+
+    try:
+        s = s.encode('ascii')
+    except UnicodeError:
+        return error()
+
+    udecode = lambda a: a.decode('utf8')
+    if s[:1] == b'.':
+        udecode = lambda a: a
+        s = s[1:]
+    s = s.replace(b'~', b'=')
+    try:
+        return udecode(b64decode(s, '-_'))
+    except Exception:
+        try:
+            # For retrocompatibility
+            return udecode(b64decode(s))
+        except Exception:
+            pass
+        return error()
+
+
```

Comment:
> I more or less understand the = to ~ replacement (though I don't understand why it's necessary; ?foo=bar== seems to survive transport just fine afaict).

I think it's just to be [safe](http://stackoverflow.com/a/941481/2643846).

> bytes/unicode conversion.

[`Base64`](https://docs.python.org/3/library/base64.html#base64.b64encode) converts *binary* to text. In python 2, a normal string `isinstance` of `bytes`, so that'll work. But in python 3, every string is a `unicode` object which is not `isinstance` of `bytes`. Hence the conversion. Related issue - [liberapay#235](https://github.com/liberapay/liberapay.com/issues/235)

> . prefix

This is a little tricky to explain. If we do encode the string to `utf-8` while encoding to `base64` as mentioned above, then, while decoding from `base64`, we want to [decode](https://github.com/gratipay/gratipay.com/pull/3965/files#diff-29dde85b4bbb7c47ef404ed0ab9f054eR160) the string from `utf-8` as well to get the original string back. *But..* if we do not encode the string to `utf-8` *and* the string had some characters which can't be decoded with `utf-8` ([try-except block](https://github.com/gratipay/gratipay.com/pull/3965/files#diff-29dde85b4bbb7c47ef404ed0ab9f054eR142)) , we [*do not*](https://github.com/gratipay/gratipay.com/pull/3965/files#diff-29dde85b4bbb7c47ef404ed0ab9f054eR162) want to apply the `utf-8` decoding while `base64` decoding. The `.` prefix is just used as a marker for that.
Not sure how much sense that makes, but if you run the encoding/decoding with `'\x80abc'` as an example, it should be clear.

> use-case for default

It's useful for providing a more helpful context specific error message. For example, instead of showing `invalid base64 input` as the error message in email verification, we let the caller handle the error through `default` and show a more helpful error message to the user. Not very relevant in this case maybe, but useful if we use encoding in future at other places as well. :)